src/riscv/lib/src/machine_state/page_cache/dispatch.rs

```diff
@@ -301,36 +320,41 @@ impl<MC: MemoryConfig + Send> Default for OutlineCompiler<MC> {
 }
 
 impl<MC: MemoryConfig + Send> DispatchCompiler<MC> for OutlineCompiler<MC> {
-    fn should_compile(&self, target: &mut DispatchTarget<Self, MC>) -> bool {
-        unsafe {
-            match target.remaining_calls {
-                0 => true,
-                _ => {
-                    //SAFETY: `remaining_calls` (a usize) can only be a positive, non-zero integer
-                    target.remaining_calls = target.remaining_calls.unchecked_sub(1);
-                    false
-                }
+    fn should_compile(&self, target: &DispatchTarget<Self, MC>) -> bool {
+        // We use *relaxed* ordering here as only the 'main' thread touches this
+        // counter. The background JIT thread does not inspect it.
+        //
+        // Since we do not require synchronisation here, Relaxed is sufficient for
+        // the single thread inspecting and updating this value.
+        //
+        // On both `x64` and `ARM-64` (the main platforms we care about), using relaxed
+        // atomic store/load produces identical code to a non-atomic 64-bit load/store.
+        //
+        // While we could use a non-atomic store/load here - using the atomic at least
+        // allows us to avoid `unsafe`, which is not required as it would perform no better.
```

Comment:
This property happens to be true right now but it is not enforced. How can it be enforced? If it's not possible, what's the impact of not relying on that property and making the atomic operation universally safe?